mautic/mautic
# jQuery and other libraries update into max minor versions

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [y]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [n]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://mautic.atlassian.net/browse/TPROD-460



#### Description:

Since we've moved most libraries into NPM we can now use it to update packages. I've learned that NPM itself does not update packages in package.json but there is https://www.npmjs.com/package/npm-check-updates for the job. So I installed it globally and run:
```
$ ncu
Checking /var/www/html/package.json
[====================] 34/34 100%

 @claviska/jquery-minicolors   ^2.2.4  →   ^2.3.6
 bootstrap                     ^3.3.7  →   ^5.2.3
 chart.js                      ^2.3.0  →   ^4.3.0
 ckeditor4                    ^4.16.0  →  ^4.21.0
 codemirror                   ^5.15.2  →   ^6.0.1
 dropzone                      ^4.0.1  →   ^5.9.3
 grunt-contrib-cssmin          ^4.0.0  →   ^5.0.0
 jquery                        ^3.3.1  →   ^3.7.0
 jquery-datetimepicker        ^2.5.20  →  ^2.5.21
 jquery-form                   ^4.2.2  →   ^4.3.0
 js-cookie                     ^2.0.2  →   ^3.0.5
 jsplumb                      ^2.12.8  →  ^2.15.6
 moment                        ^2.8.1  →  ^2.29.4
 mousetrap                     ^1.4.6  →   ^1.6.5
 shufflejs                     ^5.2.2  →   ^6.1.0
 typeahead.js                 ^0.10.2  →  ^0.11.1
```
It shows each version in color as the documentation suggests. So I run this command to update all libraries to max minor versions:
```
$ ncu --target minor -u
Upgrading /var/www/html/package.json
[====================] 34/34 100%

 @claviska/jquery-minicolors   ^2.2.4  →    ^2.3.6
 bootstrap                     ^3.3.7  →    ^3.4.1
 chart.js                      ^2.3.0  →    ^2.9.4
 ckeditor4                    ^4.16.0  →   ^4.21.0
 codemirror                   ^5.15.2  →  ^5.65.13
 dropzone                      ^4.0.1  →    ^4.3.0
 jquery                        ^3.3.1  →    ^3.7.0
 jquery-datetimepicker        ^2.5.20  →   ^2.5.21
 jquery-form                   ^4.2.2  →    ^4.3.0
 js-cookie                     ^2.0.2  →    ^2.2.1
 jsplumb                      ^2.12.8  →   ^2.15.6
 moment                        ^2.8.1  →   ^2.29.4
 mousetrap                     ^1.4.6  →    ^1.6.5
 shufflejs                     ^5.2.2  →    ^5.4.1
 typeahead.js                 ^0.10.2  →   ^0.11.1

Run npm install to install new versions.
```
That updated package.json. Then I run `npm i` to update also package.lock and actually get the latest versions of the libraries.

When i run `ncu` again we get smaller output:
```
ncu
Checking /var/www/html/package.json
[====================] 34/34 100%

 bootstrap               ^3.4.1  →  ^5.2.3
 chart.js                ^2.9.4  →  ^4.3.0
 codemirror            ^5.65.13  →  ^6.0.1
 dropzone                ^4.3.0  →  ^5.9.3
 grunt-contrib-cssmin    ^4.0.0  →  ^5.0.0
 js-cookie               ^2.2.1  →  ^3.0.5
 shufflejs               ^5.4.1  →  ^6.1.0
```
We'll have to update those libraries with more care as these would be a major release upgrades and will introduce BC changes. Coming soon. (Any help welcome)

During testing I noticed that the minicolors selector does not work properly. The background image was not showing up. I fixed it in https://github.com/mautic/mautic/commit/184578a8c427760e1ee5a6a7174a1b18f5eb1f9c

From:
```
15 vulnerabilities (1 low, 3 moderate, 6 high, 5 critical)
```
to
```
14 vulnerabilities (1 low, 2 moderate, 6 high, 5 critical)
```

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to categories and select category color.
3. Go to plugins. Resize the browser window and see that the list of plugins is changing based on the window width.
4. Visit an email or landing page detail page to see that the charts are loading correctly.
5. Upload an asset
6. Edit an email and set a publish up/down date to confirm that the date picker works.
7. Edit a campaign to ensure that the campaign builder works as before
8. Go to segments and add a segment memgership filter with including operator to ensure you can type and select the segment you want.
